common/Geometry2d/Pose.hpp

```diff
@@ -0,0 +1,193 @@
+
+    /**
+     * Operators
+     */
+    Twist operator+(const Twist& other) const {
+        return Twist(_twist + other._twist);
+    }
+    Twist operator-(const Twist& other) const {
+        return Twist(_twist - other._twist);
+    }
+    Twist operator*(double s) const { return Twist(_twist * s); }
+    Twist operator/(double s) const { return Twist(_twist / s); }
+    Twist& operator+=(const Twist& other) {
+        _twist += other._twist;
+        return *this;
+    }
+    Twist& operator-=(const Twist& other) {
+        _twist -= other._twist;
+        return *this;
+    }
+    Twist& operator*=(double s) {
+        _twist *= s;
+        return *this;
+    }
+    Twist& operator/=(double s) {
+        _twist /= s;
+        return *this;
+    }
+
+private:
+    Eigen::Vector3d _twist;
+};
+
+/**
+ * Represents a pose in 2d space: (x, y, theta). Uses double-precision floating
+ * point numbers for the coordinates, and can be used to represent the position
+ * and orientation of a robot in the plane.
+ */
+class Pose {
+public:
+    /**
+     * Default constructor - zero-initialize
+     */
+    Pose() = default;
+
+    /**
+     * Point-heading constructor
+     */
+    Pose(Point position, double heading)
+        : _pose(position.x(), position.y(), heading) {}
+
+    /**
+     * Implicit conversion from Eigen::Vector3d
+     */
+    Pose(const Eigen::Vector3d& other) : _pose(other) {}
+
+    /**
+     * Copy-constructor - default
+     */
+    Pose(const Pose& other) = default;
+
+    /**
+     * Compute the pose specified using this pose as coordinates in a frame of
+     * reference specified by `other`, in the global space.
+     *
+     * i.e.
+     *
+     * x:
+     *       |  x
+     *       | /
+     *       |/
+     * ------|-------
+     *       |
+     *       |
+     *       |
+     *
+     * y:
+     *     |
+     *     |
+     * ----|----
+     *     |\
+     *     | y
+     *
+     * y.withOrigin(x):
+     *       |  x
+     *       | / \
+     *       |/   y
+     * ------|------
+     *       |
+     */
+    Pose withOrigin(Pose other) const {
+        double sh = std::sin(other.heading()), ch = std::cos(other.heading());
+        double x = other.position().x() + ch * other.position().x() -
+                   sh * other.position().y();
+        double y =
+            other.position().y() + ch * position().y() + sh * position().x();
+        double h = heading() + other.heading();
+        return Pose({x, y}, h);
+    }
+
+    /**
+     * Implicit conversion to Eigen::Vector3d
+     */
+    operator Eigen::Vector3d() const { return _pose; }
+
+    /**
+     * Calculate a TransformMatrix corresponding to using this pose as the
+     * origin of the coordinate system.
+     */
+    TransformMatrix transform() const {
+        return TransformMatrix(position(), heading());
+    }
+
+    /**
+     * Accessors
+     */
+    Point position() const { return Point(_pose.template block<2, 1>(0, 0)); }
```

Comment:
People might expect these accessors to be references when they're using Pose btw.
```Pose pose{};
pose.position() += Point(1,1);
```
This effectively no-op code  unfortunately might or might not give a compiler warning.

Reply:
Hmm...I honestly can't immediately think of a good way to resolve this. Do you have a suggestion?

Reply:
Hmmm. You could create a Point subclass or something that is created from the 3f (with references). Or you could back Pose with Point and float instead. Idk the class idea seems similar (though different) to the idea behind MotionInstants https://github.com/RoboJackets/robocup-software/blob/master/soccer/planning/MotionInstant.hpp and RobotPose.

Reply:
Backing it with point+double gives the same issue though when you try to expose a Eigen::Vector3d and the first one is really just abusing inheritance for no good reason.  I could probably get the right semantics by creating a `PoseLinearRef` class with defined `operator+=(Point)` and implicit conversion to Point (no ref). I'll try that and see if it gets too ugly.

Reply:
Alternatively do this like ROS and express everything in terms of transforms, make `Pose` an alias for a transform, and then mutate the Pose using a transform with something like 
```c++
Pose pose{};
pose *= Transform(1, 1, 0);
```

Reply:
That's not the issue, Oswin. We have Point all over our codebase, so regardless of the "canonical" way to manipulate these at some point someone will try doing it the wrong way.

On that note, I wrote out a `PositionRef` and it's just kind of a pain to use. Unless I copy the entire Point API into the class (which I'd rather not do) it's going to not work with auto (expected, and that's honestly fine) _and_ it will be missing some of the methods that you would want to do on `Point`.

Reply:
Could you enforce that mutating these don't do anything by changing this to like a `const Point` so that the no-op code gives a compiler warning?

Reply:
Yeah that works (error not warning though, obviously). I'll do that.